### PR TITLE
Fix bugs, and enable cancellation during file sync

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#2416

### CODE REVIEW

**Comment by shuo-wu** [comment]
Will update the vendor commit when the sparse tool PR is merged

**Review by innobead — APPROVED** [review]
LGTM

Need to wait for the dependency updated before merge.

**Comment by shuo-wu** [comment]
@innobead vendor updated, with one more commit introduced. Can you take a look?

**Review by innobead — APPROVED** [review]
LGTM

File: pkg/server/downloader.go
Comment (PhanLe1010) [inline pkg/server/downloader.go:73]: IMO, we show `check and set` value for a `shared variable` while holding the lock.

For example, assume that at the beginning we have 2 goroutines (A and B) and the `d.cancelFunc == nil`. The OS schedule the goroutines in this order:
1. A enters line 66 and check d.isDownloading(). Since `d.cancelFunc == nil`, A skips to 70
2. A enters line 66 and check d.isDownloading(). Since `d.cancelFunc == nil`, B skips to 70
3. A enters line 71 and set `d.cancelFunc = cancel`
4. A processes to download the file
5. B enters line 71 and set `d.cancelFunc` to a different  `cancel` func
6. B processes to download the file
7. Now both A and B is downloading the file. Moreover, we cannot stop B because the cancel func is belongs to B

To solve this issue, we can:
d.Lock()
        if d.cancelFunc != nil {
              return fmt.Errorf("BUG: downloader gets duplicate receive requests during downloading")
        }
	d.cancelFunc = cancel
d.Unlock()

**Review by PhanLe1010 — COMMENTED** [review]
@shuo-wu  Sorry for the late response. Can you take a look?

File: pkg/server/downloader.go
Comment (shuo-wu) [reply pkg/server/downloader.go:73]: Oh, Sorry. My bad. Check & set should be atomic. Thank you for pointing out.

**Pull request merged**
